### PR TITLE
feat: Explicit buckets- add schemas on create

### LINKED ISSUE #2505: add ability to add  measurement schemas to a bucket when it is created

### PR DESCRIPTION
**Pull request by jrenee42**
Closes #2505 

Can now add measurement schemas during bucket creation.  

the measurement schema section appears/disappears when you toggle the 'schematype' radiobutton.

this means that if you add data, then switch back to implicit, the section disappears (along with its data).

it does correct validation:  it only validates the data when the schema type is explicit.
and you don't have to add measurement schemas if you are creating an explicit bucket, measurement schemas can be added later.

added these two events:

bucket.schema.explicit.creation.uploadSchema: triggers for each schema being added
bucket.creation: when a bucket is created (this didn't already exist, so added it)

Screen Shot 2021-10-05 at 2 56 30 PM

Screen Shot 2021-10-05 at 2 56 39 PM

the validation, backend calls, etc are all the same as from when the bucket already exists.

the backend call needs the bucket id, so the bucket is created first; then the schema is added.  so the notifications show up for adding the schema (successfully or unsuccessfully) too.

### CODE REVIEW

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 59 | :red_circle:
accessibility | 96 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 100 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=2810)

File: src/buckets/actions/thunks.ts [hunk 1]
```diff
@@ -144,8 +145,25 @@ export const createBucketAndUpdate = (
       bucketSchema
     )
 
+    // just created the bucket, now add any schemas that might be there:
     dispatch(addBucket(newBucket))
     dispatch(checkBucketLimits())
+
+    if (resp.data.schemaType === 'explicit' && schemas && schemas.length) {
+      // in case they choose explicit, add schemas, then change their mind back and
+      // change it to implicit; or if explicit and choose not to add schemas at this time.
```
Comment (hoorayimhelping) [inline src/buckets/actions/thunks.ts:154]: I'm a little confused by this comment. what line does it apply to? Is it guarding against the situation described in the comment, or is it enabling the situation described in the comment?

File: src/buckets/actions/thunks.ts [hunk 2]
```diff
@@ -344,6 +363,21 @@ export const addSchemaToBucket = (
   }
 }
 
+export const addSchemaToBucket = (
+  bucketID: string,
+  orgID: string,
+  bucketName: string,
+  schema: typeof MeasurementSchemaCreateRequest
+) => async (dispatch: Dispatch<Action>) => {
+  await addMeasurementSchemaToBucketInternal(
```
Comment (hoorayimhelping) [inline src/buckets/actions/thunks.ts:372]: does this need to be `await`ed if there's nothing below this code doing anything with the result?

is this equivalent to

```ts
addMeasurementSchemaToBucketInternal(
  bucketID,
  schema,
  orgID,
  bucketName,
  dispatch
).then()
```
?

File: src/buckets/components/createBucketForm/BucketOverlayForm.tsx [hunk 3]
```diff
@@ -117,30 +117,36 @@ export default class BucketOverlayForm extends PureComponent<Props> {
       />
     )
 
-    const measurementSchemaSection =
-      schemaType === 'explicit' ? measurementSchemaComponent : null
+    const showMeasurementSchemaSection =
+      (isEditing && readOnlySchemaType === 'explicit') ||
+      schemaType === 'explicit'
+
+    const measurementSchemaSection = showMeasurementSchemaSection
```
Comment (hoorayimhelping) [inline src/buckets/components/createBucketForm/BucketOverlayForm.tsx:124]: :+1: clean!

File: src/buckets/components/createBucketForm/CreateBucketForm.tsx [hunk 4]
```diff
@@ -19,16 +19,21 @@ import {
   DEFAULT_RULES,
 } from 'src/buckets/reducers/createBucket'
 import {AppState, Bucket, RetentionRule} from 'src/types'
+import {event as influxEvent} from '../../../cloud/utils/reporting'
```
Comment (hoorayimhelping) [inline src/buckets/components/createBucketForm/CreateBucketForm.tsx:22]: is `event` already used here? any reason not to just leave this as `event`?

also, can you change this import to be absolute, not relative please?

File: src/buckets/components/createBucketForm/MeasurementSchemaUtils.ts [hunk 5]
```diff
@@ -82,3 +82,17 @@ export const isNameValid = name => {
...
+    Array.isArray(newMeasurementSchemaRequests) &&
+    newMeasurementSchemaRequests.length
+
+  if (!haveSchemas) {
+    // no schemas, nothing to validate, so everything is fine
+    // even if it is in explicit mode, can add schemas later
+    return true
+  }
+  // if so, are they all valid?
+  return newMeasurementSchemaRequests.every(schema => schema.valid)
```
Comment (hoorayimhelping) [inline src/buckets/components/createBucketForm/MeasurementSchemaUtils.ts:97]: :+1: very comprehensible function

File: src/buckets/components/createBucketForm/UpdateBucketOverlay.tsx [hunk 6]
```diff
@@ -39,6 +39,7 @@ import {getBucketFailed} from 'src/shared/copy/notifications'
 import {OwnBucket} from 'src/types'
 import {CLOUD} from 'src/shared/constants'
 import {event} from '../../../cloud/utils/reporting'
+import {areNewSchemasValid} from './MeasurementSchemaUtils'
```
Comment (hoorayimhelping) [inline src/buckets/components/createBucketForm/UpdateBucketOverlay.tsx:42]: can you turn these into absolute imports please?

**Review by hoorayimhelping — COMMENTED** [review]
I have one or two requests and a couple of questions so I'm just going to leave this as a comment rather than approval or denial - let me know if the questions make sense

File: src/buckets/components/createBucketForm/CreateBucketForm.tsx (hunk 4 above)
Comment (jrenee42) [reply src/buckets/components/createBucketForm/CreateBucketForm.tsx:22]: when i put it in as 'event' i got errors.  (there are form events and such on the page).  I can try again.

will find an absolute import and copy it.

File: src/buckets/components/createBucketForm/MeasurementSchemaUtils.ts (hunk 5 above)
Comment (jrenee42) [reply src/buckets/components/createBucketForm/MeasurementSchemaUtils.ts:97]: lol :)

File: src/buckets/components/createBucketForm/CreateBucketForm.tsx (hunk 4 above)
Comment (jrenee42) [reply src/buckets/components/createBucketForm/CreateBucketForm.tsx:22]: fixed; the method that event(<event string>) gets called in has an argument/parameter called 'event'.  so fixed now.

File: src/buckets/actions/thunks.ts (hunk 2 above)
Comment (jrenee42) [reply src/buckets/actions/thunks.ts:372]: hmm...I thought I needed the await when i did the refactor.  turns out I don't.  removing it.

**Pull request merged**